fontra2fontir/src/source.rs

```diff
@@ -1,22 +1,146 @@
+        // This file is tool generated so it shouldn't be full of human error. Fail if we don't understand.
+        for (i, line) in BufReader::new(file).lines().enumerate().skip(1) {
+            let line = line.map_err(Error::IoError)?;
+            let parts: Vec<_> = line.split(';').collect();
+            if parts.len() != 2 {
+                return Err(Error::ParseError(
+                    self.glyphinfo_file.clone(),
+                    format!("Expected two parts in line {i} separated by ;"),
+                ));
+            }
+            let glyph_name = GlyphName::new(parts[0].trim());
+            // TODO: support multiple codepoints
+            let codepoint = parts[1].trim();
+            let codepoint = if !codepoint.is_empty() {
+                let Some(codepoint) = codepoint.strip_prefix("U+") else {
+                    return Err(Error::ParseError(
+                        self.glyphinfo_file.clone(),
+                        format!("Unintelligible codepoint at line {i}"),
+                    ));
+                };
+                Some(u32::from_str_radix(codepoint, 16).map_err(|e| {
+                    Error::ParseError(
+                        self.glyphinfo_file.clone(),
+                        format!("Unintelligible codepoint at line {i}: {e}"),
+                    )
+                })?)
+            } else {
+                None
+            };
+            let glyph_file = self.glyph_file(glyph_name.clone());
+            if !glyph_file.is_file() {
+                return Err(Error::FileExpected(glyph_file));
+            }
+            if self
+                .glyph_info
+                .insert(glyph_name.clone(), (glyph_file, codepoint))
+                .is_some()
+            {
+                return Err(Error::ParseError(
+                    self.glyphinfo_file.clone(),
+                    format!("Multiple definitions of '{glyph_name}'"),
+                ));
+            }
+        }
+        Ok(())
+    }
+
+    fn glyph_state(&mut self) -> Result<HashMap<GlyphName, StateSet>, Error> {
+        let mut glyph_state = HashMap::default();
+        for (glyph_name, (glyph_file, _)) in self.glyph_info.iter() {
+            let mut tracker = StateSet::new();
+            tracker.track_file(glyph_file)?;
+            if glyph_state.insert(glyph_name.clone(), tracker).is_some() {
+                return Err(Error::ParseError(
+                    self.glyphinfo_file.clone(),
+                    format!("Multiple definitions of '{glyph_name}'"),
+                ));
+            }
+        }
+        Ok(glyph_state)
     }
 }
 
 impl Source for FontraIrSource {
     fn inputs(&mut self) -> Result<fontir::source::Input, fontir::error::Error> {
-        todo!()
+        let static_metadata = self.static_metadata_state()?;
+        self.load_glyphinfo()?;
+        let glyphs = self.glyph_state()?;
+
+        // Not sure how fontra features work
```

Comment:
I think fontra doesn't support OpenType Features yet

Reply:
I suppose it's coming? @justvanrossum do you know?

Reply:
We have no concrete timeline for that.